### PR TITLE
*: Implement PerformanceSchema with MemoryTable

### PR DESCRIPTION
**Pull request by shenli**
@nieyy @coocood

### CODE REVIEW

File: optimizer/optimizer.go [hunk 1]
```diff
@@ -69,7 +69,7 @@ func (c *supportChecker) Enter(in ast.Node) (ast.Node, bool) {
 	switch x := in.(type) {
 	case *ast.TableName:
 		if strings.EqualFold(x.Schema.O, perfschema.Name) {
-			c.unsupported = true
+			c.unsupported = false
```
Comment (nieyy) [inline optimizer/optimizer.go]: Why not just remove this `case`?

File: perfschema/perfschema.go [hunk 2]
```diff
@@ -33,40 +31,69 @@ type StatementInstrument interface {
...
-// PerfHandle is the only access point for the in-memory performance schema information
-var PerfHandle PerfSchema
+// NewPerfHandle creates a new perfSchema on store.
+func NewPerfHandle(store kv.Storage) PerfSchema {
+	schema := PerfHandle.(*perfSchema)
+	schema.store = store
+	schema.historyHandles = make([]int64, 0, stmtsHistoryElemMax)
+	_ = schema.initialize()
+	// Existing instrument names are the same as MySQL 5.7
+	PerfHandle.RegisterStatement("sql", "alter_table", (*ast.AlterTableStmt)(nil))
```
Comment (nieyy) [inline perfschema/perfschema.go]: I would prefer move this `RegisterStatement` calls to `statement.go` since it's all related to statement instruments.

**Comment by nieyy** [comment]
Looks like we don't need `perfschema/common.go` any more.

File: perfschema/perfschema.go (hunk 2 above)
Comment (shenli) [inline perfschema/perfschema.go]: It should be called in this function. Maybe I can put them in a function in statement.go.

File: optimizer/optimizer.go (hunk 1 above)
Comment (shenli) [inline optimizer/optimizer.go]: I will remove it.

**Comment by shenli** [comment]
@nieyy PTAL

File: perfschema/perfschema.go [hunk 3]
```diff
@@ -33,40 +30,45 @@ type StatementInstrument interface {
...
-	fields map[string][]*field.ResultField
-	stores map[string]*leveldb.DB
-	lsns   map[string]*uint64
+	store   kv.Storage
+	dbInfo  *model.DBInfo
+	tables  map[string]*model.TableInfo
+	mTables map[string]table.Table // MemoryTables for perfSchema
+
+	// Used for TableStmtsHistory
+	historyHandles []int64
```
Comment (nieyy) [inline perfschema/perfschema.go]: `events_statements_current` table only maintains the latest instrument record for each connection, that is why I use `connID` as key, here I did not see how to ensure that, could you plz show me?

**Comment by nieyy** [comment]
Also, setup_xxx tables and events_xxx tables need to restrict access permissions:
1. All tables support `SELECT` operation

Besides that:
1. `events_xxx` tables support `TRUNCATE` only
2. `setup_timers`, `setup_consumers` and `setup_instruments` support `UPDATE` only
3. `setup_actors`, `setup_objects` support `INSERT`, `DELETE` and `TRUNCATE` only

For operations not supported, an error should be prompted to user.

File: perfschema/perfschema.go (hunk 3 above)
Comment (shenli) [inline perfschema/perfschema.go]: First at all, I use the THREAD_ID column as PK for the table. See: https://github.com/pingcap/tidb/blob/shenli/perfoschema/perfschema/init.go#L233https://github.com/pingcap/tidb/blob/shenli/perfoschema/perfschema/init.go#L233
So its value will be used as handle.  See: https://github.com/pingcap/tidb/blob/shenli/perfoschema/table/tables/memory_tables.go#L188

When insert data failed for KeyExistsErr, I will try to update the record. See:  https://github.com/pingcap/tidb/blob/shenli/perfoschema/perfschema/statement.go#L286

Comment (shenli) [inline perfschema/perfschema.go]: There is a little problem about connID. If you access TiDB from tidb-server, there will be a connID. If you access TiDB from sql-driver, the connID will always be zero.

**Comment by shenli** [comment]
@nieyy We have a privilege system. So we can grant privilege for those tables. Maybe in the latter PR.

**Comment by nieyy** [comment]
@shenli If the change is not too much, can we finish in this PR? at least we can restrict all tables support `SELECT` only first to avoid faulty operations.

**Comment by coocood** [comment]
@nieyy @shenli 

I think privilege should be added in future PR, because the purpose of this PR is refactoring, privilege feature is not implemented in the old code. 

Normally refactoring should not add new features.

**Comment by coocood** [comment]
LGTM

File: infoschema/infoschema_test.go [hunk 4]
```diff
@@ -101,13 +102,13 @@ func (*testSuite) TestT(c *C) {
 	is := handle.Get()
 
 	schemaNames := is.AllSchemaNames()
-	c.Assert(len(schemaNames), Equals, 2)
-	c.Assert(testutil.CompareUnorderedStringSlice(schemaNames, []string{infoschema.Name, "Test"}), IsTrue)
+	c.Assert(len(schemaNames), Equals, 3)
```
Comment (qiuyesuifeng) [inline infoschema/infoschema_test.go]: HasLen

File: perfschema/init.go [hunk 5]
```diff
@@ -192,20 +192,86 @@ var stagesCurrentCols = []columnInfo{
...
+func createMemoryTable(meta *model.TableInfo, alloc autoid.Allocator) (table.Table, error) {
+	tbl, _ := tables.MemoryTableFromMeta(alloc, meta)
+	return tbl, nil
+}
+
+func (ps *perfSchema) buildTables() error {
+	tbls := make([]*model.TableInfo, 0, len(ps.tables))
+	ps.mTables = make(map[string]table.Table, len(ps.tables))
+	var (
+		err error
```
Comment (qiuyesuifeng) [inline perfschema/init.go]: Seems no need to define err?

File: perfschema/init.go [hunk 6]
```diff
@@ -192,20 +192,86 @@ var stagesCurrentCols = []columnInfo{
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func genGlobalID(store kv.Storage) (int64, error) {
+	var globalID int64
+	err := kv.RunInNewTxn(store, true, func(txn kv.Transaction) error {
```
Comment (qiuyesuifeng) [inline perfschema/init.go]: Why generate every id in a new txn? Seems in setColumnID function, one txn is enough?

Comment (shenli) [inline perfschema/init.go]: All id generating in a table use the same key. So we use a new txn to make this process short. And the SQL txn will be less possibly to conflict.

**Comment by shenli** [comment]
@qiuyesuifeng  PTAL

**Comment by nieyy** [comment]
@shenli If the privilege system feature is not implemented yet, I agree it can be implemented in the future PR.

LGTM

**Pull request merged**